<a href="https://colab.research.google.com/github/lych0508/Projects/blob/main/Copy_of_GoHealthUrgentCare_Assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

## Background

You are on the People Analytics Team and are asked to help us understand certification compliance. 

You have data from a few places: 

> Employee Data - contains employee demographics; 


> certificationd data - contains data on certifications; 


> certification classifications - contains data on how to group certifications and roles they are applicable to.


## Objective










1.   Understand which employees are out of compliance (i.e. have expired certifications)
2.   Sliced into each feature on segmentation analysis

## Importing Libraries

### Libraries

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from datetime import date

### Loading data

In [2]:
import io
import pandas as pd
from google.colab import files

In [3]:
uploaded = files.upload()

Saving People_Analytics.xlsx to People_Analytics.xlsx


In [4]:
xlsx_file = io.BytesIO(uploaded.get('People_Analytics.xlsx'))

In [5]:
df = pd.read_excel(xlsx_file, 'Sheet1')

In [6]:
df.head()

Employee Number              Certification Certification Date Issued  \
0           100040         Basic Life Support                2022-08-16   
1           100040  Cert Prov Order Entry Pro                2022-08-12   
2           100071    Natl Certified Med Asst                2022-03-17   
3           100080         Basic Life Support                2021-12-29   
4           100080    Natl Certified Med Asst                2022-06-29   

        Effective End Certification Status Hire Date Rehire Date  \
0 2023-02-16 23:59:00               Active       NaT         NaN   
1 2023-08-12 23:59:00               Active       NaT         NaN   
2 2023-03-17 23:59:00               Active       NaT         NaN   
3 2023-12-29 23:59:00               Active       NaT         NaN   
4 2023-07-14 23:59:00               Active       NaT         NaN   

  Employee Start Date Pay Class Job Name Job Family Name Market  \
0                 NaT       NaN      NaN             NaN    NaN   
1                 NaT       NaN      NaN             NaN    NaN   
2                 NaT       NaN      NaN             NaN    NaN   
3                 NaT       NaN      NaN             NaN    NaN   
4                 NaT       NaN      NaN             NaN    NaN   

  Source of hire Current Employee Rehired Employee Cert Status  \
0            NaN               No     NaN               Active   
1            NaN               No     NaN               Active   
2            NaN               No     NaN               Active   
3            NaN               No     NaN               Active   
4            NaN               No     NaN               Active   

  Certification type Medical Assistant Requirement Rad Tech Requirement  \
0          First Aid                             Y                    Y   
1     Medical Scribe                             N                    N   
2  Medical Assistant                             Y                    N   
3          First Aid                             Y                    Y   
4  Medical Assistant                             Y                    N   

  Medical Scribe Requirement  
0                          Y  
1                          Y  
2                          N  
3                          Y  
4                          N

#Exploratory data analysis

First, I check if the dataset has any the missing values and data type for furture analysis

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2690 entries, 0 to 2689
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Employee Number                2690 non-null   int64         
 1   Certification                  2502 non-null   object        
 2   Certification Date Issued      2502 non-null   datetime64[ns]
 3   Effective End                  2502 non-null   datetime64[ns]
 4   Certification Status           2690 non-null   object        
 5   Hire Date                      1219 non-null   datetime64[ns]
 6   Rehire Date                    1219 non-null   object        
 7   Employee Start Date            1219 non-null   datetime64[ns]
 8   Pay Class                      1219 non-null   object        
 9   Job Name                       1219 non-null   object        
 10  Job Family Name                1219 non-null   object        
 11  Market           

In [8]:
#employee number should be string
#rehire date should be date 
df['Employee Number'] = df['Employee Number'].astype(str)
df['Rehire Date']= pd.to_datetime(df['Rehire Date'], errors='coerce')


After join these three dataset, I found there are missing values in hiring date. I believe these employess are not working at the company at more. So I will filter them out

In [9]:

df = df.loc[df['Hire Date'].notnull()]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1219 entries, 1471 to 2689
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Employee Number                1219 non-null   object        
 1   Certification                  1031 non-null   object        
 2   Certification Date Issued      1031 non-null   datetime64[ns]
 3   Effective End                  1031 non-null   datetime64[ns]
 4   Certification Status           1219 non-null   object        
 5   Hire Date                      1219 non-null   datetime64[ns]
 6   Rehire Date                    1218 non-null   datetime64[ns]
 7   Employee Start Date            1219 non-null   datetime64[ns]
 8   Pay Class                      1219 non-null   object        
 9   Job Name                       1219 non-null   object        
 10  Job Family Name                1219 non-null   object        
 11  Market        

I assume there are some employees who don't have certifications, therefore there are some missing values in certification column. 

In [10]:
# filter employee with certification
df1 = df.loc[df['Certification'].notnull()]
df1.isnull().sum()

Employee Number                  0
Certification                    0
Certification Date Issued        0
Effective End                    0
Certification Status             0
Hire Date                        0
Rehire Date                      1
Employee Start Date              0
Pay Class                        0
Job Name                         0
Job Family Name                  0
Market                           0
Source of hire                   0
Current Employee                 0
Rehired                          0
Employee Cert Status             0
Certification type               0
Medical Assistant Requirement    0
Rad Tech Requirement             0
Medical Scribe Requirement       0
dtype: int64

In [11]:

df_nan = df1[df1['Rehire Date'].isnull()]
df_nan 

Employee Number Certification Certification Date Issued  \
2454          227496        No BLS                2022-07-05   

           Effective End Certification Status  Hire Date Rehire Date  \
2454 2023-01-05 23:59:00               Active 2022-07-05         NaT   

     Employee Start Date  Pay Class           Job Name    Job Family Name  \
2454          2022-07-05  Temporary  Medical Assistant  Medical Assistant   

        Market Source of hire Current Employee  Rehired Employee Cert Status  \
2454  Market C         Indeed              Yes  Rehired               Active   

     Certification type Medical Assistant Requirement Rad Tech Requirement  \
2454          First Aid                             Y                    Y   

     Medical Scribe Requirement  
2454                          Y

In [12]:
from pandas._libs import NaT
# I go back to chekc the dataset, the rehire date = hire date
df1['Rehire Date'] = df1['Rehire Date'].replace(NaT, '2022-07-05')
df1.isnull().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Employee Number                  0
Certification                    0
Certification Date Issued        0
Effective End                    0
Certification Status             0
Hire Date                        0
Rehire Date                      0
Employee Start Date              0
Pay Class                        0
Job Name                         0
Job Family Name                  0
Market                           0
Source of hire                   0
Current Employee                 0
Rehired                          0
Employee Cert Status             0
Certification type               0
Medical Assistant Requirement    0
Rad Tech Requirement             0
Medical Scribe Requirement       0
dtype: int64

In [13]:
#number of expired cert
expired_df = df1.loc[df1['Certification Status'] == 'Expired']
expired_df

Employee Number            Certification Certification Date Issued  \
1474          105125       Basic Life Support                2020-10-22   
1480          106012      CA Rad Tech License                2020-01-20   
1519          106160            R.T.(R)(ARRT)                2021-10-12   
1525          106171            R.T.(R)(ARRT)                2021-11-10   
1542          109013            R.T.(R)(ARRT)                2021-10-05   
...              ...                      ...                       ...   
2360          227092                   No BLS                2022-05-14   
2361          227099                   No BLS                2022-05-09   
2362          227099          Non-Certifed MA                2022-05-09   
2363          227109                   No BLS                2022-05-09   
2573          227782  Natl Certified Med Asst                2003-10-22   

           Effective End Certification Status  Hire Date Rehire Date  \
1474 2022-10-22 23:59:00              Expired 2017-01-09  2017-01-09   
1480 2022-10-31 23:59:00              Expired 2017-01-04  2017-01-04   
1519 2022-10-30 23:59:00              Expired 2019-06-24  2019-06-24   
1525 2022-10-31 23:59:00              Expired 2019-09-16  2019-09-16   
1542 2022-09-30 23:59:00              Expired 2017-06-26  2017-06-26   
...                  ...                  ...        ...         ...   
2360 2022-11-14 23:59:00              Expired 2022-05-14  2022-05-14   
2361 2022-11-09 23:59:00              Expired 2022-05-09  2022-05-09   
2362 2022-11-09 23:59:00              Expired 2022-05-09  2022-05-09   
2363 2022-11-09 23:59:00              Expired 2022-05-16  2022-05-16   
2573 2022-09-15 23:59:00              Expired 2022-08-02  2022-08-02   

     Employee Start Date Pay Class                        Job Name  \
1474          2017-01-09        FT  Radiologic Technologist (Flex)   
1480          2017-01-04        FT         Radiologic Technologist   
1519          2019-06-24        FT         Radiologic Technologist   
1525          2019-09-16        FT  Radiologic Technologist (Flex)   
1542          2017-06-26        FT         Radiologic Technologist   
...                  ...       ...                             ...   
2360          2022-05-14        FT               Medical Assistant   
2361          2022-05-09  Per Diem               Medical Assistant   
2362          2022-05-09  Per Diem               Medical Assistant   
2363          2022-05-16        FT               Medical Assistant   
2573          2022-08-02        FT       Medical Assistant (Float)   

        Job Family Name    Market       Source of hire Current Employee  \
1474           Rad Tech  Market B               Indeed              Yes   
1480           Rad Tech  Market A               Indeed              Yes   
1519           Rad Tech  Market A               Indeed              Yes   
1525           Rad Tech  Market A               Indeed              Yes   
1542           Rad Tech  Market B             Job Fair              Yes   
...                 ...       ...                  ...              ...   
2360  Medical Assistant  Market A               Indeed              Yes   
2361  Medical Assistant  Market D             Job Fair              Yes   
2362  Medical Assistant  Market D             Job Fair              Yes   
2363  Medical Assistant  Market E               Indeed              Yes   
2573  Medical Assistant  Market C  ClinicalCareers.com              Yes   

          Rehired Employee Cert Status Certification type  \
1474  Non-Rehired              Expired          First Aid   
1480  Non-Rehired              Expired           Rad Tech   
1519  Non-Rehired              Expired           Rad Tech   
1525  Non-Rehired              Expired           Rad Tech   
1542  Non-Rehired              Expired           Rad Tech   
...           ...                  ...                ...   
2360  Non-Rehired              Expired          First Aid   
2361  Non

In [14]:
expired_df['Employee Number'].unique()

array(['105125', '106012', '106160', '106171', '109013', '109014',
       '109024', '109034', '109075', '109098', '109127', '109138',
       '109149', '111018', '111103', '111124', '111130', '200336',
       '200510', '200849', '201522', '201569', '201729', '201840',
       '201939', '202704', '202769', '202859', '203148', '203306',
       '203319', '203490', '203777', '204056', '204298', '204320',
       '204325', '204524', '214948', '215500', '215501', '215551',
       '216307', '216373', '216393', '216669', '216680', '216700',
       '216718', '216737', '216748', '216760', '216776', '216816',
       '216818', '216849', '216897', '216904', '216906', '216907',
       '216914', '216916', '216917', '216920', '226969', '226973',
       '226974', '226987', '227007', '227009', '227011', '227019',
       '227020', '227041', '227084', '227085', '227090', '227092',
       '227099', '227109', '227782'], dtype=object)

In [15]:
# employees who has more than one expired certification
expired_df['Employee Number'].value_counts()[expired_df['Employee Number'].value_counts()>1]

215501    2
216904    2
226974    2
216669    2
216920    2
216917    2
216916    2
216914    2
216906    2
216849    2
200849    2
216818    2
216816    2
204320    2
204524    2
215500    2
216737    2
216718    2
226987    2
226973    2
227007    2
109138    2
227099    2
109014    2
227090    2
227085    2
109075    2
227009    2
109098    2
109149    2
227041    2
227019    2
Name: Employee Number, dtype: int64

#Feature Engineering

In [16]:
# Drop some useless columns
to_drop = ['Employee Number','Certification Date Issued','Effective End','Hire Date', 'Rehire Date', 'Employee Start Date', 'Employee Cert Status']
df2 = df1.drop(to_drop, axis=1)


In [17]:
# yes and no have to be converted to boolean values
df2["Current Employee"] = df2['Current Employee'] == 'Yes'
df2["Medical Assistant Requirement"] = df2['Medical Assistant Requirement'] == 'Y'
df2["Medical Scribe Requirement"] = df2['Medical Scribe Requirement'] == 'Y'
df2["Rad Tech Requirement"] = df2['Rad Tech Requirement'] == 'Y'
df2["Certification Status"] = df2['Certification Status'] == 'Active'
df2["Rehired"] = df2['Rehired'] == 'Rehired'
df2.head()


Certification  Certification Status Pay Class  \
1471   Basic Life Support                  True        FT   
1472  CA Rad Tech License                  True        FT   
1473        R.T.(R)(ARRT)                  True        FT   
1474   Basic Life Support                 False        FT   
1475  CT Rad Tech License                  True        FT   

                                   Job Name Job Family Name    Market  \
1471  Team Leader (Radiologic Technologist)        Rad Tech  Market A   
1472  Team Leader (Radiologic Technologist)        Rad Tech  Market A   
1473  Team Leader (Radiologic Technologist)        Rad Tech  Market A   
1474         Radiologic Technologist (Flex)        Rad Tech  Market B   
1475         Radiologic Technologist (Flex)        Rad Tech  Market B   

     Source of hire  Current Employee  Rehired Certification type  \
1471       Job Fair              True    False          First Aid   
1472       Job Fair              True    False           Rad Tech   
1473       Job Fair              True    False           Rad Tech   
1474         Indeed              True    False          First Aid   
1475         Indeed              True    False           Rad Tech   

      Medical Assistant Requirement  Rad Tech Requirement  \
1471                           True                  True   
1472                          False                  True   
1473                          False                  True   
1474                           True                  True   
1475                          False                  True   

      Medical Scribe Requirement  
1471                        True  
1472                       False  
1473                       False  
1474                        True  
1475                       False

In [18]:
df2[['Pay Class']].value_counts()

Pay Class
FT           641
Per Diem     272
Temporary     75
PT            42
Seasonal       1
dtype: int64

In [19]:
replace_values = {'PT' : "Others", 'Seasonal' : 'Others' } 
df2 = df2.replace({"Pay Class": replace_values})        

In [20]:
df2[['Job Name']].value_counts()

Job Name                                           
Medical Assistant                                      366
Radiologic Technologist                                282
Medical Assistant (Float)                              127
Advanced Clinical Associate                            117
Radiologic Technologist (Flex)                          27
Radiologic Technologist (Float)                         26
Medical Assistant - Limited X-Ray                       23
Team Leader (Radiologic Technologist)                   23
Team Leader (Medical Assistant)                         21
Medical Scribe                                           5
Lead Medical Assistant                                   3
Learning Coach (Medical Assistant)                       3
Quality Assurance & Control Radiologic Technologist      3
Lead Medical Assistant (DTE)                             2
Lead Radiologic Technologist                             2
Medical Assistant (DTE)                                  1
dtyp

In [21]:
replace_values = {'Medical Scribe': 'Others', 'Lead Medical Assistant' : 'Others', 'Learning Coach (Medical Assistant)' : 'Others', 'Quality Assurance & Control Radiologic Technologist' : 'Others', 
                  'Lead Medical Assistant (DTE)' : 'Others', 'Lead Radiologic Technologist' : 'Others', 'Medical Assistant (DTE)' : 'Others'} 
df2 = df2.replace({"Job Name": replace_values}) 

#Model

In [22]:
X = df2.drop('Certification Status', 1)
y = df2['Certification Status']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  """Entry point for launching an IPython kernel.


In [23]:
X = pd.get_dummies(X)


In [24]:
from sklearn.model_selection import train_test_split
X_train, X_cv, y_train, y_cv = train_test_split(X, y, test_size=0.3)

In [25]:
from sklearn.tree import DecisionTreeClassifier

In [26]:
model = DecisionTreeClassifier(random_state=1)

In [27]:
model.fit(X_train,y_train)

DecisionTreeClassifier(random_state=1)

In [28]:
predictions = model.predict(X_cv)

In [29]:
from sklearn.metrics import accuracy_score, classification_report

In [30]:
print(classification_report(y_cv, predictions))

              precision    recall  f1-score   support

       False       0.16      0.19      0.18        31
        True       0.91      0.89      0.90       279

    accuracy                           0.82       310
   macro avg       0.54      0.54      0.54       310
weighted avg       0.83      0.82      0.83       310



In [31]:
print(accuracy_score(y_cv, predictions))

0.8193548387096774


In [32]:
from sklearn.ensemble import RandomForestClassifier

In [33]:
rfmodel = RandomForestClassifier(n_estimators=500)

In [34]:
rfmodel.fit(X_train, y_train)

RandomForestClassifier(n_estimators=500)

In [35]:
rfpredictions = rfmodel.predict(X_cv)

In [36]:
print(accuracy_score(y_cv, rfpredictions))

0.8741935483870967


In [37]:
print(classification_report(y_cv, rfpredictions))

              precision    recall  f1-score   support

       False       0.17      0.06      0.09        31
        True       0.90      0.96      0.93       279

    accuracy                           0.87       310
   macro avg       0.53      0.51      0.51       310
weighted avg       0.83      0.87      0.85       310



# Feature Importance

In [38]:
# check feature importance of random forest for feature selection
forest = rfmodel.fit(X_train, y_train)

importances = forest.feature_importances_

# Print the feature ranking
print("Feature importance ranking by Random Forest Model:")
for k,v in sorted(zip(map(lambda x: round(x, 4), importances), X.columns), reverse=True):
    print (v + ": " + str(k))

Feature importance ranking by Random Forest Model:
Rehired: 0.074
Source of hire_Indeed: 0.0562
Pay Class_FT: 0.0506
Source of hire_ClinicalCareers.com: 0.0488
Market_Market B: 0.048
Market_Market E: 0.0479
Certification_No BLS: 0.0477
Pay Class_Per Diem: 0.0461
Market_Market A: 0.0447
Market_Market D: 0.042
Source of hire_Career site: 0.0379
Market_Market C: 0.0326
Certification_Basic Life Support: 0.0292
Job Name_Medical Assistant: 0.0262
Certification_Non-Certifed MA: 0.025
Source of hire_Recruiter: 0.024
Job Name_Radiologic Technologist: 0.0234
Job Name_Medical Assistant - Limited X-Ray: 0.022
Source of hire_Job Fair: 0.021
Source of hire_Referral: 0.0194
Pay Class_Temporary: 0.0176
Job Name_Medical Assistant (Float): 0.0159
Certification_R.T.(R)(ARRT): 0.0128
Job Family Name_Medical Assistant: 0.0116
Job Name_Radiologic Technologist (Flex): 0.0112
Job Family Name_Rad Tech: 0.0103
Job Name_Others: 0.01
Certification_Certified Medical Assistant: 0.0098
Certification type_First Aid: 

#Summary


- 11% of total employee have expired certifications
- In general, the Job function, Market, Pay class and Sources of hire with the most employees will have the most expired certificates
- Medical Assistant has highest number of expired certifications because it requires 14 certificates whereas Rad Tech and Medical Scribe only require 10 certificates each
- BLS, Non-Certified MA and R.T.(R)(ARRT) are most likely to expire and they are needed by both Medical Assistant and Rad Tech
- Market D has high expired certification because it only hires Medical Assistants
- Rehired employees follow overall employee trend for certification compliance as it has 14% expiration rate whereas the overall company has 11%
- Feature importance ranking using Random Forest matches our result, model accuracy 87%

